src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -28,75 +28,102 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
-    const auto input_size = static_cast<int>(values.size() * sizeof(T));
-    auto output_size = LZ4_compressBound(input_size);
+    /**
+     * Use the LZ4 high compression API to compress the copied values. As C-library LZ4 needs raw pointers as input
+     * and output. To avoid directly handling raw pointers we use std::vectors as input and output. The input vector
+     * contains the data that needs to be compressed and the output vector is allocated enough memory to contain
+     * the compression result. Via the .data() call we can supply LZ4 with raw pointers to the memory the vectors use.
+     * These are cast to char-pointers since LZ4 expects char pointers.
+     */
+    DebugAssert(values.size() * sizeof(T) <= std::numeric_limits<int>::max(),
+                "Input of LZ4 encoder contains too many bytes to fit into a 32-bit signed integer sized vector that is"
+                " used by the LZ4 library.");
+
+    const auto input_size = values.size() * sizeof(T);
+    // estimate the (maximum) output size
+    const auto output_size = LZ4_compressBound(static_cast<int>(input_size));
     auto compressed_data = pmr_vector<char>{alloc};
     compressed_data.resize(static_cast<size_t>(output_size));
     const int compression_result = LZ4_compress_HC(reinterpret_cast<char*>(values.data()), compressed_data.data(),
-                                                   input_size, output_size, LZ4HC_CLEVEL_MAX);
+                                                   static_cast<int>(input_size), output_size, LZ4HC_CLEVEL_MAX);
     Assert(compression_result > 0, "LZ4 compression failed");
 
-    auto data_ptr = std::allocate_shared<pmr_vector<char>>(alloc, std::move(compressed_data));
-    auto null_values_ptr = std::allocate_shared<pmr_vector<bool>>(alloc, std::move(null_values));
+    // shrink the vector to the actual size of the compressed result
+    compressed_data.resize(static_cast<size_t>(compression_result));
+    compressed_data.shrink_to_fit();
 
-    return std::allocate_shared<LZ4Segment<T>>(alloc, data_ptr, null_values_ptr, nullptr, compression_result,
-                                               input_size, num_elements);
+    return std::allocate_shared<LZ4Segment<T>>(alloc, std::move(compressed_data), std::move(null_values), nullptr,
+                                               input_size);
   }
 
   std::shared_ptr<BaseEncodedSegment> _on_encode(
       const std::shared_ptr<const ValueSegment<std::string>>& value_segment) {
     const auto alloc = value_segment->values().get_allocator();
     const auto num_elements = value_segment->size();
-    DebugAssert(num_elements <= std::numeric_limits<int>::max(),
-                "Trying to compress a ValueSegment with more "
-                "elements than fit into an int.");
+
+    /**
+     * First iterate over the values for two reasons.
+     * 1) If all the strings are empty LZ4 will try to compress an empty vector which will cause a segmentation fault.
```

Comment:
What about a vector with only empty strings except for one single cell, which contains exactly one char? Does LZ4 have something like a minimum data size which it needs in order to work?

Reply:
I added an extra test case for that. A single char works.